libraries/botframework-config/src/botConfiguration.ts

```diff
@@ -0,0 +1,230 @@
+        let bot = BotConfiguration.fromJSON(JSON.parse(await txtfile.read(botpath)));
+        bot.internal.location = botpath;
+
+        let hasSecret = (bot.secretKey && bot.secretKey.length > 0);
+        if (hasSecret)
+            bot.decrypt(secret);
+
+        return bot;
+    }
+
+    // save the config file
+    public async save(botpath?: string, secret?: string): Promise<void> {
+        if (!!secret) {
+            this.validateSecretKey(secret);
+        }
+
+        let hasSecret = (this.secretKey && this.secretKey.length > 0);
+
+        // make sure that all dispatch serviceIds still match services that are in the bot
+        for (let service of this.services) {
+            if (service.type == ServiceTypes.Dispatch) {
+                let dispatchService = <IDispatchService>service;
+                dispatchService.serviceIds = Enumerable.fromSource(dispatchService.serviceIds)
+                    .where(serviceId => Enumerable.fromSource(this.services).any(s => s.id == serviceId))
+                    .toArray();
+            }
+        }
+
+        if (hasSecret)
+            this.encrypt(secret);
+
+        await fsx.writeJson(botpath || <string>this.internal.location, this.toJSON(), { spaces: 4 });
+
+        if (hasSecret)
+            this.decrypt(secret);
+    }
+
+    public clearSecret() {
+        this.secretKey = '';
+    }
+
+    // connect to a service
+    public connectService(newService: IConnectedService): void {
+        if (Enumerable.fromSource(this.services)
+            .where(s => s.type == newService.type)
+            .where(s => s.id == newService.id)
+            .any()) {
+            throw Error(`service with ${newService.id} already connected`);
+        } else {
+            // give unique name
+            let nameCount = 1;
+            let name = newService.name;
+
+            while (true) {
+                if (nameCount > 1) {
+                    name = `${newService.name} (${nameCount})`;
+                }
+
+                if (!Enumerable.fromSource(this.services).where(s => s.name == name).any())
+                    break;
+                nameCount++;
+            }
+            newService.name = name;
+
+            this.services.push(BotConfiguration.serviceFromJSON(newService));
+        }
+    }
+
+    // encrypt all values in the config
+    public encrypt(secret: string) {
+        this.validateSecretKey(secret);
+
+        for (let service of this.services) {
+            (<ConnectedService>service).encrypt(secret);
+        }
+    }
+
+    // decrypt all values in the config
+    public decrypt(secret?: string) {
+        this.validateSecretKey(secret);
+
+        for (let service of this.services) {
+            (<ConnectedService>service).decrypt(secret);
+        }
+    }
+
+
+    // remove service by name or id
+    public disconnectServiceByNameOrId(nameOrId: string): IConnectedService {
+        let svs = new List<IConnectedService>(this.services);
```

Comment:
This can be simplified by the following:
```js
const {services = []} = this;
let i = services.length;
while (i--) {
  const service = services[i];
 if (service.id == nameOrId || service.name == nameOrId) {
    return services.splice(i, 1)[0];
  }
}
```
It also does not require taking on a dependency.